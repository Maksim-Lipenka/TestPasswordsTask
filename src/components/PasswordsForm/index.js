import React from 'react';
import {
  StyleSheet,
  Text,
  View,
  SafeAreaView,
  ScrollView,
  TextInput,
  TouchableOpacity,
} from 'react-native';
import {connect} from 'react-redux';
import {compose} from 'redux';
import {Field, reduxForm, formValueSelector} from 'redux-form';
import {InputText, Button} from '../core';
import {atLeastOneCapitalLetter, atLeastOneNumber, atLeastOneRegularLetter, atLeastSixCharacters, required, } from '../../validation';
import { form, passwordConfirmationField, passwordConfirmationLabel, passwordField, passwordLabel, passwordRuleCapitalLetter, passwordRuleLength, passwordRuleMatches, passwordRuleNumber, passwordRuleRegularLetter, passwordRulesTitle, submitButtonLabel } from './strings';

const passwordsMatch = (value, allValues) =>
  value !== allValues.password ? "Passwords don't match" : undefined;

const PasswordsForm = props => {
  const {handleSubmit, passwordValue, passwordConfirmationValue} = props;

  const submit = values => {
    alert('Congratulations!');
  };

  return (
    <SafeAreaView style={styles.wrapper}>
      <View style={styles.container}>
        <Field
          name={passwordField}
          component={InputText}
          label={passwordLabel}
          validate={[required, atLeastOneRegularLetter, atLeastOneCapitalLetter, atLeastOneNumber, atLeastSixCharacters]}
          showIndicator
          secureTextEntry
        />
        <Field
          name={passwordConfirmationField}
          component={InputText}
          validate={[required, passwordsMatch]}
          label={passwordConfirmationLabel}
          showIndicator
          secureTextEntry
        />
        <View>
          <Text>{passwordRulesTitle}</Text>
          <Text style={[styles.indicatorNotValid, passwordValue && !atLeastOneRegularLetter(passwordValue) && styles.indicatorValid]}>{passwordRuleRegularLetter}</Text>
          <Text style={[styles.indicatorNotValid, passwordValue && !atLeastOneCapitalLetter(passwordValue) && styles.indicatorValid]}>{passwordRuleCapitalLetter}</Text>
          <Text style={[styles.indicatorNotValid, passwordValue && !atLeastOneNumber(passwordValue) && styles.indicatorValid]}>{passwordRuleNumber}</Text>
          <Text style={[styles.indicatorNotValid, passwordValue && !atLeastSixCharacters(passwordValue) && styles.indicatorValid]}>{passwordRuleLength}</Text>
          <Text style={[styles.indicatorNotValid, passwordValue && !passwordsMatch(passwordConfirmationValue, { password: passwordValue }) && styles.indicatorValid]}>{passwordRuleMatches}</Text>
        </View>
        <Button
          onPress={handleSubmit(submit)}
          label={submitButtonLabel}
          disabled={!props.valid}
        />
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  wrapper: {
    alignItems: 'center',
  },
  container: {
    height: '100%',
    width: 250,
    justifyContent: 'center',
    alignItems: 'center',
  },
  indicatorNotValid: {
    color: 'red'
  },
  indicatorValid: {
    color: 'green'
  }
});

export default compose(
  reduxForm({
    form: form,
  }),
  connect(state => ({
    passwordValue: formValueSelector(form)(state, passwordField),
    passwordConfirmationValue: formValueSelector(form)(state, passwordConfirmationField),
  })),
)(PasswordsForm);
