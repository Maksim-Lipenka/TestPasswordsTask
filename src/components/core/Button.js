import React from 'react'
import {Text, TouchableOpacity, StyleSheet} from 'react-native'

export default ({onPress, label="Submit", disabled}) => {
  return (
  <TouchableOpacity
    disabled={disabled}
    style={[styles.submitButton, disabled && styles.submitButtonDisabled]}
    onPress={onPress}>
    <Text style={styles.submitButtonTitle}><Text>{label}</Text></Text>
  </TouchableOpacity>
  )
}

const styles = StyleSheet.create({
  submitButton: {
    backgroundColor: '#3fd1a5',
    marginTop: 10,
    textAlign: 'center',
    width: '100%',
    height: 37,
    justifyContent: 'center',
    alignItems: 'center',
  },
  submitButtonDisabled: {
    backgroundColor: 'gray'
  },
  submitButtonTitle: {
    color: 'white',
    fontWeight: 'bold',
  },
});