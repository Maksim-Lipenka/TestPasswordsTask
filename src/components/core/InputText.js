import React from 'react'
import {View, Text, TextInput, StyleSheet} from 'react-native'

export default ({input: {onChange, ...restInput}, meta: { error }, label, showIndicator, inputContainerStyle, inputStyle, ...props}) => {
  
  return (
    <View style={[styles.inputWrapper, inputContainerStyle]}>
      <Text>{label}</Text>
      <TextInput
        style={[styles.input, inputStyle]}
        onChangeText={onChange}
        autoCapitalize='none'
        {...restInput}
        {...props}
      />
      {showIndicator && <View style={[styles.indicator, { backgroundColor: error ? "red" : "green" }]}/>}
    </View>
  );
};

const styles = StyleSheet.create({
  inputWrapper: {
    width: '100%'
  },
  input: {
    borderColor: '#1f1f1f',
    borderWidth: 1,
    height: 37,
    width: '100%',
    marginBottom: 12,
    paddingLeft: 8,
    paddingRight: 32
  },
  indicator: {
    position: 'absolute',
    right: 12,
    bottom: 49/2+3,
    width: 6,
    height: 6,
    borderRadius: 50
  }
});