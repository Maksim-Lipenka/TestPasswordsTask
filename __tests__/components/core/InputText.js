import 'react-native';
import React from 'react';
import {InputText} from '../../../src/components/core';
import renderer from 'react-test-renderer';

describe('InputText', () => {
  const handlePress = jest.fn();

  const defaultProps = {
    input: {onChange: handlePress},
    meta: {error: false},
    label: "Password",
    showIndicator: false,
  };

  it('renders correctly', () => {
    const tree = renderer.create(<InputText {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
