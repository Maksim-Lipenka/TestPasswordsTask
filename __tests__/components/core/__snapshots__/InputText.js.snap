// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputText renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <Text>
    Password
  </Text>
  <TextInput
    autoCapitalize="none"
    onChangeText={[MockFunction]}
    style={
      Array [
        Object {
          "borderColor": "#1f1f1f",
          "borderWidth": 1,
          "height": 37,
          "marginBottom": 12,
          "paddingLeft": 8,
          "paddingRight": 32,
          "width": "100%",
        },
        undefined,
      ]
    }
  />
</View>
`;
